mhs-reference-implementation/mhs/SDSLookup/sds.py

```diff
@@ -0,0 +1,105 @@
+CA_CERTS_FILE = Path(ROOT_DIR) / 'data' / 'certs' / 'client.cert'
+
+
+class SDSClient:
+    mhs_attributes = [
+        'nhsEPInteractionType', 'nhsIDCode', 'nhsMhsCPAId', 'nhsMHSEndPoint', 'nhsMhsFQDN',
+        'nhsMHsIN', 'nhsMHSIsAuthenticated', 'nhsMHSPartyKey', 'nhsMHsSN', 'nhsMhsSvcIA', 'nhsProductKey',
+        'uniqueIdentifier', 'nhsMHSAckRequested', 'nhsMHSActor', 'nhsMHSDuplicateElimination',
+        'nhsMHSPersistDuration', 'nhsMHSRetries', 'nhsMHSRetryInterval', 'nhsMHSSyncReplyMode'
+    ]
+
+    def __init__(self, ldap_address: str = None, tls: bool = False):
+        if ldap_address:
+            if tls:
+                load_tls = ldap3.Tls(local_private_key_file=PRIVATE_KEY,
+                                     local_certificate_file=LOCAL_CERT_FILE,
+                                     validate=ssl.CERT_REQUIRED, version=ssl.PROTOCOL_TLSv1,
+                                     ca_certs_file=CA_CERTS_FILE)
+
+                self.server = ldap3.Server(ldap_address, use_ssl=True, tls=load_tls)
+            else:
+                self.server = ldap3.Server(ldap_address)
+
+            self.connection = ldap3.Connection(self.server, auto_bind=True, client_strategy=ldap3.REUSABLE)
+        else:
+            self.server = None
+            self.connection = None
+
+    async def get_mhs_details(self, nhs_id_code: str, service_id: str):
+        """
+        Returns the all end point details for the given org code and service id
+        :param nhs_id_code:
+        :param service_id:
+        :return:
+        """
+
+        accredited_system_lookup = await self._accredited_system_lookup(nhs_id_code, service_id)
+
+        if not accredited_system_lookup:
+            raise Exception('No response from accredited system lookup')
+
+        # As per the spec exactly one result should be returned, manage error checking here
+        response = accredited_system_lookup[0]
+        party_key = response['attributes']['nhsMHSPartyKey']
+
+        if party_key:
+            details = await self._mhs_details_lookup(party_key, service_id)
+            return details[0]
+
+    async def _accredited_system_lookup(self, org_code: str, service_id: str):
+        """
+            Used to find an accredited system, the result contains the ASID and nhsMhsPartyKey.
+            These can then be used to find an MHS endpoint
+        :return: endpoint details - filtered to only contain nhsMHSPartyKey
+        """
+        if not self.connection:
+            raise RuntimeError('No connection established')
+
+        search_filter = f"(&(nhsIDCode={org_code}) (objectClass={AS_OBJECT_CLASS}) (nhsAsSvcIA={service_id}))"
+
+        message_id = self.connection.search(search_base=NHS_SERVICES_BASE,
+                                            search_filter=search_filter,
+                                            attributes="nhsMHSPartyKey")
+
+        task = IOLoop.current().run_in_executor(None, self._get_query_result, message_id)
```

Comment:
It might be worth extracting this to a 'run task asynchronously' function, as we might end up using it elsewhere, and it would also serve to explain what you're doing a bit more.

Also, I appreciate that this came from the example I provided, but I'm wondering if we could do this without needing a Tornado class? I think the Tornado functions are now a pretty thin wrapper around the built-in async IO functionality, so we should be able to do the same thing with asyncio. This package otherwise doesn't need to know anything about Tornado.

Reply:
I've swapped the tornado call to use the native async/await calls and as a result I've held off from moving the call into another method

Reply:
As discussed I've wrapped the call within the `get_query_result` method to use the asyncio `run_in_executor` method